### ISSUE TITLE
PPO - invalid action masking

### ISSUE THREAD

**Issue by roeslib** [aai-institute/tianshou#887]
Hi there! Base on the atari_ppo.py example I'm implementing my PPO model. 
I need to work with an invalid action masking which affects the distribution probability, each observation generates an invalid action masking that affects the distribution probability to sample only allowed actions. 
Is an example with invalid action masking implemented? or where and how in the code can I add an invalid action masking in the example code of PPO that evaluates each observation?

**Comment by spacegoing** [comment]
@roeslib I do not quite understand your question. Are you suggesting the policy's distribution depends on observations (e.g., a categorical distribution with various categories depending on obs)? Or truncating actions after being sampled from policy?

For the latter case, you can implement a customized `map_action()` https://github.com/thu-ml/tianshou/blob/f8808d236f53d86583c74f459365c9cc4184256c/tianshou/policy/base.py#L153

**Comment by roeslib** [comment]
Thanks for your answer what I am looking for is an invalid action masking, the actor generates logits and those logits needs to be masked through an "invalid action mask" that sets 0 to the logits that are not valid to be choosen when sampling, next come the softmax that returns the distribution probability to start sampling actions. An example code:

Part of code of actor.forward:
```
        location_masks = []
        for obs in batch.obs:
            box_mask = get_rotation_mask(obs, (10,10,10))

        location_masks = torch.FloatTensor(np.array(location_masks)).to("cuda")

        logits, hidden = self.actor(batch.obs, state=state, info=batch.info)

        dist = self.dist_fn(logits,location_masks)

        act = dist.sample()
            
        return Batch(logits=logits, act=act, state=hidden, dist=dist)
```

code of dist:
```
    FixedCategorical = torch.distributions.Categorical

    old_sample = FixedCategorical.sample
    FixedCategorical.sample = lambda self: old_sample(self).unsqueeze(-1)

    log_prob_cat = FixedCategorical.log_prob
    FixedCategorical.log_prob = lambda self, actions: log_prob_cat(
    self, actions.squeeze(-1)).view(actions.size(0), -1).sum(-1).unsqueeze(-1)

    FixedCategorical.mode = lambda self: self.probs.argmax(dim=-1, keepdim=True)
           
    def dist(logits, location_masks):
        ones = torch.ones_like(location_masks)
        inver_mask = ones - location_masks

        lx = F.softmax(logits - inver_mask * 14, dim=-1)
        lx = lx + 1e-5

        fat_cat = FixedCategorical(probs=lx)
        return fat_cat
```


Apart from this, in my "info" (retun of step from gym) there is a key name "ratio" that saves the ratio of filling of my bin, I need to see its episodic information mean and std. By now in the log messages in the terminal I see reward and losses I would like to add the episodic "ratio", what should I do to get in the log messages in terminal the "ratio" episodic values?

**Comment by spacegoing** [comment]
Hi Rose, I'm not on the dev team. Just try to help here:

For Q1 action map:

I'm still not sure I completely understand your question. But for policy-specific information, there is a reserved key called `policy` in `ReplayBuffer`. and you may store & retrieve your action masks through `batch.policy.action_mask`.
https://github.com/thu-ml/tianshou/blob/f8808d236f53d86583c74f459365c9cc4184256c/tianshou/data/buffer/base.py#L33

For episodic info:
You may implement a customized `collector`:
https://github.com/thu-ml/tianshou/blob/f8808d236f53d86583c74f459365c9cc4184256c/tianshou/data/collector.py#L366
https://github.com/thu-ml/tianshou/blob/f8808d236f53d86583c74f459365c9cc4184256c/tianshou/data/collector.py#L398

**Comment by roeslib** [comment]
Thank you for your answer, I will give it try but perhaps you have already done it and you could show me an example how you implemented a costumized Collector.

**Comment by spacegoing** [comment]
@roeslib Hi Rose, unfortunately, regret to tell you that I currently do not have an example that would directly address your requirement. I'm only a tianshou user, not on the dev team:P

As someone who frequently navigates the project's codebase, I generally have a sense of where to make modifications to meet specific needs. But in this case, I'm unable to provide a detailed solution to your specific situation.

I'd also like to kindly remind you that the issues section on GitHub is primarily intended for discussions related to the open source project itself, such as bug reports, feature requests, or general questions about the project's usage. It's not typically used for addressing individual, personalized requirements.

For questions about how to implement specific functionality in a more personalized context, platforms like StackOverflow or Reddit might be better suited. These platforms have a broader user base and are designed to handle a wide variety of programming-related questions and discussions.

Thank you for your understanding and best of luck with your project!

**Comment by MischaPanch** [comment]
I would close this since it's really not an issue of tianshou. @roeslib generally tianshou aims at being useful for research, but your use-case is very specific, so I don't think it makes sense to discuss how it can best be supported in an issue
